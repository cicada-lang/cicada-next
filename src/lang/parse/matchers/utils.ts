import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
