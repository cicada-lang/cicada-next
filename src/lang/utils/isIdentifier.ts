const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
