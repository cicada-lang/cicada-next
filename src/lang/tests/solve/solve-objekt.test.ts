import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("solve Objekt", async () => {
  const output = await runCode(`

class ABC {
  a: String
  b: String
  c: String
}

solve (a: String, b: String, c: String) {
  unify { a, b, c } = { a: "a", b: "b", c: "c" } : ABC
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ a: \\"a\\", b: \\"b\\", c: \\"c\\" }"',
  )
})

test("solve Objekt -- occur twice", async () => {
  const output = await runCode(`

class ABC {
  a: String
  b: String
  c: String
}

solve (a: String, b: String) {
  unify { a, b, c: b } = { a: b, b: "c", c: "c" } : ABC
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: \\"c\\", b: \\"c\\" }"')
})

test("solve Objekt -- extra common properties", async () => {
  await expectCodeToFail(`

class AB {
  a: String
  b: String
}

solve (a: String, b: String, c: String) {
  unify { a, b, c } = { a: b, b: "c", c: "c" } : AB
}

`)
})
