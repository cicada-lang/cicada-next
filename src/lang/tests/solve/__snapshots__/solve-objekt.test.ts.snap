// Vitest Snapshot v1, https://bun:test.dev/guide/snapshot.html

exports[`solve Objekt -- deepWalk 1`] = `
"{
  a: "a",
  b: "b",
  c: "c",
  abc: {
    a: "a",
    b: "b",
    c: "c"
  }
}"
`;

exports[`solve Objekt -- occur twice 1`] = `
"{
  a: "c",
  b: "c"
}"
`;

exports[`solve Objekt 1`] = `
"{
  a: "a",
  b: "b",
  c: "c"
}"
`;
